// adwavecss/dist/index.js
export class Theme {
    static dark = "dark-theme";
    static light = "light-theme";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/box.md
 */
export class Box {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/box.md#ClassNames
     */
    static box = "box";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/box.md#ClassNames
     */
    static rounded = "rounded";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/box.md#ClassNames
     */
    static bg2 = "bg-level-2";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/box.md#ClassNames
     */
    static bg3 = "bg-level-3";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/box.md#ClassNames
     */
    static bg4 = "bg-level-4";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md
 */
export class Button {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static button = "btn";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static disabled = "disabled";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static primary = "primary";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static danger = "danger";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static flat = "flat";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static pill = "pill";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static circular = "circular";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static square = "square";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static toggled = "toggled";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static wrapper = "btn-wrapper";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/button.md#ClassNames
     */
    static adaptive = "adaptive";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/linked.md
     */
    static linked = "linked";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/linked.md
     */
    static linkedVertical = "vertical";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/card.md
 */
export class Card {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/card.md#ClassNames
     */
    static card = "card";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/card.md#ClassNames
     */
    static activable = "activable";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/checkbox.md
 */
export class Checkbox {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/checkbox.md#ClassNames
     */
    static checkbox = "checkbox";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/checkbox.md#ClassNames
     */
    static disabled = "disabled";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/frame.md
 */
export class Frame {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/frame.md#ClassNames
     */
    static frame = "frame";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/input.md
 */
export class Input {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/input.md#ClassNames
     */
    static input = "input";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/input.md#ClassNames
     */
    static disabled = "disabled";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/linked.md
     */
    static linked = "linked";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/linked.md
     */
    static linkedVertical = "vertical";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/input.md#ClassNames
     */
    static wrapper = "input-wrapper";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/list.md
 */
export class List {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/list.md#ClassNames
     */
    static list = "list";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/list.md#ClassNames
     */
    static element = "list-element";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/list.md#ClassNames
     */
    static activableElement = "activable";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/message.md
 */
export class Message {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/message.md#ClassNames
     */
    static message = "message";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/message.md#ClassNames
     */
    static success = "success";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/message.md#ClassNames
     */
    static alert = "alert";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/message.md#ClassNames
     */
    static error = "error";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/nav-sidebar.md
 */
export class NavSidebar {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/nav-sidebar.md#ClassNames
     */
    static navSidebar = "nav-sidebar";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/nav-sidebar.md#ClassNames
     */
    static button = "nav-sidebar-btn";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/nav-sidebar.md#ClassNames
     */
    static separator = "separator";
    static active = "active";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/slider.md
 */
export class Slider {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/slider.md#ClassNames
     */
    static slider = "slider";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/slider.md#ClassNames
     */
    static disabled = "disabled";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/slider.md#ClassNames
     */
    static track = "slider-track";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/slider.md#ClassNames
     */
    static progress = "slider-progress";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/slider.md#ClassNames
     */
    static thumb = "slider-thumb";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/slider.md#ClassNames
     */
    static noPosition = "no-position";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/switch.md
 */
export class Switch {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/switch.md#ClassNames
     */
    static switch = "switch";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/switch.md#ClassNames
     */
    static disabled = "disabled";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/switch.md#ClassNames
     */
    static knob = "switch-knob";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/switch.md#ClassNames
     */
    static active = "active";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/typography.md
 */
export class Typography {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/typography.md#ClassNames
     */
    static text = "text";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/typography.md#ClassNames
     */
    static subtitle = "subtitle";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/typography.md#ClassNames
     */
    static label = "label";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/typography.md#ClassNames
     */
    static header = "header";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/suggestions.md
 */
export class Suggestions {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/suggestions.md#ClassNames
     */
    static suggestions = "suggestions";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/suggestions.md#ClassNames
     */
    static option = "suggestions-option";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/suggestions.md#ClassNames
     */
    static active = "active";
}
/**
 * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md
 */
export class Selector {
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static selector = "selector";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static disabled = "disabled";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static selectedOption = "selected-option";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static downButton = "down-button";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static opened = "opened";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static optionsList = "options-list";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static option = "option";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static top = "top";
    /**
     * Read more at @link https://github.com/ncpa0/ADWaveCSS/blob/master/docs/components/selector.md#ClassNames
     */
    static noPosition = "no-position";
}
Object.freeze(Theme);
Object.freeze(Box);
Object.freeze(Button);
Object.freeze(Card);
Object.freeze(Checkbox);
Object.freeze(Frame);
Object.freeze(Input);
Object.freeze(List);
Object.freeze(Message);
Object.freeze(NavSidebar);
Object.freeze(Slider);
Object.freeze(Switch);
Object.freeze(Typography);
Object.freeze(Suggestions);
Object.freeze(Selector);
export default {
    Theme,
    Box,
    Button,
    Card,
    Checkbox,
    Frame,
    Input,
    List,
    Message,
    NavSidebar,
    Slider,
    Switch,
    Typography,
    Suggestions,
    Selector,
};
