// jsxte/dist/legacy/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  DefaultTemplateArrayCache: () => import_default_cache.DefaultTemplateArrayCache,
  ErrorBoundary: () => import_error_boundary.ErrorBoundary,
  Interpolate: () => import_interpolate.Interpolate,
  InterpolateTag: () => import_interpolate.InterpolateTag,
  JsxteRenderError: () => import_jsxte_render_error.JsxteRenderError,
  createElement: () => import_jsx_runtime.createElement,
  defineContext: () => import_component_api.defineContext,
  memo: () => import_memo.memo,
  renderToHtml: () => import_render_to_html.renderToHtml,
  renderToHtmlAsync: () => import_render_to_html.renderToHtmlAsync,
  renderToJson: () => import_render_to_json.renderToJson,
  renderToJsonAsync: () => import_render_to_json.renderToJsonAsync,
  renderToStringTemplateTag: () => import_render_to_string_template_tag.renderToStringTemplateTag
});
module.exports = __toCommonJS(src_exports);
var import_array_flat_polyfill = require("./utilities/array-flat-polyfill.js");
__reExport(src_exports, require("./express/index.js"), module.exports);
__reExport(src_exports, require("./jsx/jsx.types.js"), module.exports);
var import_interpolate = require("./string-template-renderer/interpolate.js");
var import_default_cache = require("./string-template-renderer/default-cache.js");
var import_error_boundary = require("./error-boundary/error-boundary.js");
var import_component_api = require("./component-api/component-api.js");
var import_render_to_html = require("./html-renderer/render-to-html.js");
var import_render_to_json = require("./json-renderer/render-to-json.js");
var import_render_to_string_template_tag = require("./string-template-renderer/render-to-string-template-tag.js");
var import_memo = require("./utilities/memo.js");
var import_jsx_runtime = require("./jsx/jsx-runtime.js");
var import_jsxte_render_error = require("./jsxte-render-error.js");
