// jsxte/dist/cjs/utilities/cache.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utilities/cache.ts
var cache_exports = {};
__export(cache_exports, {
  Cache: () => Cache
});
module.exports = __toCommonJS(cache_exports);
var CacheEntry = class {
  constructor(props, value, maxAge) {
    this.value = value;
    this.maxAge = maxAge;
    this.createdAt = Date.now();
    this.props = new Map(Object.entries(props));
  }
  isExpired(now) {
    return now - this.createdAt > this.maxAge;
  }
  compareProps(propEntries) {
    if (propEntries.length !== this.props.size)
      return false;
    for (let i = 0; i < propEntries.length; i++) {
      const [key, value] = propEntries[i];
      const cachedPropValue = this.props.get(key);
      if (cachedPropValue !== value)
        return false;
    }
    return true;
  }
};
var Cache = class {
  constructor(maxAge = 15 * 60 * 1e3, maxEntries = 10) {
    this.maxAge = maxAge;
    this.maxEntries = maxEntries;
    this.lastInvalidate = Date.now();
    this.invalidationInterval = 5 * 60 * 1e3;
    this.entries = [];
    if (maxEntries < 1) {
      throw new Error("Cache maxEntries must be greater than 0");
    }
  }
  enforceEntryLimit() {
    while (this.entries.length > this.maxEntries) {
      this.entries.shift();
    }
  }
  invalidateExpired() {
    const now = Date.now();
    this.entries = this.entries.filter((entry) => !entry.isExpired(now));
  }
  get(props) {
    if (Date.now() - this.lastInvalidate > this.invalidationInterval) {
      this.lastInvalidate = Date.now();
      setTimeout(() => this.invalidateExpired(), 0);
    }
    const propsEntries = Array.from(Object.entries(props));
    const cacheEntry = this.entries.find(
      (entry) => entry.compareProps(propsEntries)
    );
    if (cacheEntry && !cacheEntry.isExpired(Date.now()))
      return cacheEntry.value;
    return void 0;
  }
  set(props, value) {
    this.entries.push(new CacheEntry(props, value, this.maxAge));
    setTimeout(() => this.enforceEntryLimit(), 0);
  }
};
